### PR TITLE
server: expose buffered bytes from Acceptor

### PR DESCRIPTION
**Pull request by djc**
As a potential help for #1915. Not sure if this is the right solution but seems like a low-cost API to offer.

### CODE REVIEW

File: rustls/src/server/server_conn.rs [hunk 1]
```diff
@@ -799,6 +799,21 @@ mod connection {
...
+        /// Yields an iterator over the buffered chunks of bytes
+        ///
+        /// Will yield an empty iterator after `accept()` has returned `Ok(Some(_))` or `Err(_)`.
+        pub fn buffered(&self) -> impl Iterator<Item = &[u8]> {
+            self.inner
+                .iter()
+                .map(|conn| {
+                    conn.core
+                        .common_state
+                        .sendable_tls
```
Comment (ctz) [inline rustls/src/server/server_conn.rs:812]: Is this the right object? If I understood the purpose of this correctly, shouldn't this be `deframer_buffer`?

Comment (djc) [reply rustls/src/server/server_conn.rs]: 🤦 You're right, of course. Still pretty straightforward...

File: rustls/src/server/server_conn.rs [hunk 2]
```diff
@@ -799,6 +799,16 @@ mod connection {
                 sig_schemes,
             }))
         }
+
+        /// Yields an iterator over the buffered chunks of bytes
+        ///
+        /// Will yield an empty iterator after `accept()` has returned `Ok(Some(_))` or `Err(_)`.
```
Comment (cpu) [inline rustls/src/server/server_conn.rs:805]: I think this is returning a slice or an empty slice in the current impl, not an iterator. Does this bit of docs need updating?

**Comment by cpu** [comment]
If the goal is to make something like https://github.com/rustls/tokio-rustls/issues/54 easier I don't know that this meaningfully shifts the needle while also introducing more API surface on a core type. 

Maybe it should be something higher-level in the acceptor API specifically for the "is this actually a plaintext connection I'm about to accept?" case?

File: rustls/src/server/server_conn.rs (hunk 2 above)
Comment (djc) [reply rustls/src/server/server_conn.rs:805]: Yup, but let's first decide if we think this API is useful.

**Comment by djc** [comment]
> If the goal is to make something like [rustls/tokio-rustls#54](https://github.com/rustls/tokio-rustls/issues/54) easier I don't know that this meaningfully shifts the needle.
> 
> Maybe it should be something higher-level in the acceptor API specifically for the "is this actually a plaintext connection I'm about to accept?" case?

That seems reasonable. So what should that API look like? Maybe `take_plaintext() -> Option<Vec<u8>>`, which can be called before `accept()` and does minimal checking if the buffered bytes look like TLS?

**Comment by cpu** [comment]
> Maybe take_plaintext() -> Option<Vec<u8>>, which can be called before accept() and does minimal checking if the buffered bytes look like TLS?

Yeah! Admittedly I haven't tried sketching anything out myself but this is more along the lines of what I was thinking.

**Comment by cpu** [comment]
@djc WDYT about something [like this approach](https://github.com/rustls/rustls/compare/main...cpu:rustls:cpu-acceptor-plaintext?expand=1) as an alternative?

I tried to come up with an adjustment that wouldn't break semver, so it's a little bit more awkward than it could be if we're open to breaking changes in this area. For e.g. having to wrap the underlying `Error` in an `OtherError` and make callers downcast to interrogate could be avoided by making a new error return for the `accept()` fn, something like `AcceptError(Error, Vec<u8>)`.

**Comment by cpu** [comment]
I think based on the conclusions in #1915 this won't be necessary. I'm going to close it for now.

**Pull request closed without merging**